File: cachesim.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int debug = 1;

char memory[1 << 24]; 

int log2(int n) {
    int r = 0;
    while (n >>= 1) r++;
    return r;
}

struct block {
    int valid;
    int dirty;
    int tag;
    char data[1024];
};

int main(int argc, char* argv[]) {
    // ./cachesim <trace-file> <cache-size-kB> <associativity> <block-size>
    // ./cachesim traces/example.txt 1024 4 32
    FILE* file = fopen(argv[1], "r");
    int cache_size = atoi(argv[2]) * 1024; // 1 KB = 1024 bytes in CS 250
    int num_ways = atoi(argv[3]);
    int block_size = atoi(argv[4]);

    // Calculations for cache
    int num_frames = cache_size / block_size;
    int num_sets = num_frames / num_ways;
    int offset_bits = log2(block_size);
    int index_bits = log2(num_sets);
    int tag_bits = 24 - offset_bits - index_bits;

    int index_mask = (1 << index_bits) - 1;
    int offset_mask = (1 << offset_bits) - 1;

    // Cache = 2D array
    struct block** cache = (struct block**)malloc(num_sets * sizeof(struct block*));
    for (int i = 0; i < num_sets; i++) {
        cache[i] = (struct block*)malloc(num_ways * sizeof(struct block));
    }
    for (int i = 0; i < num_sets; i++) { // Initialize 
        for (int j = 0; j < num_ways; j++) {
            cache[i][j].valid = 0;
            cache[i][j].dirty = 0;
            cache[i][j].tag = 0;
            memset(cache[i][j].data, 0, sizeof(cache[i][j].data)); // IS THIS OKAY
        }
    }

    // Initialize
    char type[8]; // 5 (max) + 1 (null) = 6
    int address; // 24-bit address, in hex 
    int access_size; // in bytes, no larger than 8 at a time
    char data[block_size]; // since we aren't writing across blocks

    while (fscanf(file, "%s 0x%x %d", type, &address, &access_size) != EOF) {
        // if (debug) printf("%s 0x%x %d", type, address, access_size);

        int tag = address >> (offset_bits + index_bits);
        int index = (address >> offset_bits) & index_mask;
        int block_offset = address & offset_mask;
        
        int hit = 0;
        int hit_idx = 0;

        // LOAD
        if (strcmp(type, "load") == 0) {
            for (int j = 0; j < num_ways; j++) {
                if (cache[index][j].tag == tag && cache[index][j].valid) {
                    hit = 1;
                    hit_idx = j; // Locate cache block to use
                    break;
                }
            }

            // HIT
            if (hit == 1) {
                printf("%s 0x%x hit ", type, address);
                for (int i = 0; i < access_size; i++) { // Value loaded from cache/memory
                    printf("%02hhx", cache[index][hit_idx].data[block_offset + i]); // Starting position in block, iterate over all bytes
                }
                printf("\n");
            }
            
            // MISS
            else { // UPDATE 0 INDEX AFTER IMPLMENTING LRU
                for (int i = 0; i < block_size; i++) { // Write previous data back to lower memory (entire block of data)
                    cache[index][0].data[i] = memory[address + i];
                }
                cache[index][0].tag = tag; // Which memory block cache block corresponds to
                cache[index][0].valid = 1; // Mark as not dirty
                
                printf("%s 0x%x miss ", type, address); // Return data
                for (int i = 0; i < access_size; i++) {
                    printf("%02hhx", cache[index][0].data[block_offset + i]);
                }
                printf("\n");
            } // Note: Beginning of address til end of access size but bring entire block in
        }

        // STORE
        if (strcmp(type, "store") == 0) {
            for (int i = 0; i < access_size; i++) { // Read data value to be written, store in data array
                fscanf(file, "%2hhx", &data[i]);
            }
            
            for (int j = 0; j < num_ways; j++) {
                if (cache[index][j].tag == tag && cache[index][j].valid) {
                    hit = 1;
                    hit_idx = j; // Locate cache block to use
                    break;
                }
            }

            // HIT
            if (hit == 1) {
                printf("%s 0x%x hit\n", type, address);

                for (int i = 0; i < access_size; i++) { // Write new data into cache block
                    cache[index][hit_idx].data[block_offset + i] = data[i];
                }
                
            }

            // MISS
            else {
                for (int i = 0; i < access_size; i++) { // Write previous data back to lower memory (accessed data)
                    memory[address + i] = data[i];
                }

                for (int i = 0; i < access_size; i++) { // Write new data into cache block
                    cache[index][0].data[block_offset + i] = data[i];
                }
                printf("%s 0x%x miss\n", type, address);
            }

            cache[index][hit_idx].dirty = 1; // Mark as dirty
    }
    for (int i = 0; i < num_sets; i++) {
        free(cache[i]);
    }
    free(cache);

    fclose(file);
    return EXIT_SUCCESS;
    }
}

File: cachesim.java
import java.util.*;                
import java.io.*;
import java.nio.ByteBuffer;

public class cachesim{
    public static CacheSet[] cache; // array of cache sets (2D array)
    public static byte[] mainMemory = new byte[16*1024*1024];
    public static int currentTime = 0;
    public static int blockSize; // in bytes
    public static int cacheSize; // in KB
    public static int associativity;
    public static int numSets;

    public class CacheBlock {
        boolean valid;
        boolean dirty;
        long tag;
        byte[] data;
        int lastUsedTime;
        int frameStart;
        public CacheBlock (int blockSize) {
            this.valid = false;
            this.dirty = false;
            this.tag = -1;
            this.frameStart = 0;
            this.data = new byte[blockSize];
            this.lastUsedTime = 0;
        }
    }

    public class CacheSet {
        CacheBlock[] blocks; // one set of blocks

        public CacheSet(int associativity, int blockSize) {
            blocks = new CacheBlock[associativity]; // array is of associativity length
            for (int i = 0; i < associativity; i++) { 
                blocks[i] = new CacheBlock(blockSize);
            }
        }

        public CacheBlock findLRU() {
        CacheBlock lru = blocks[0];
            for (CacheBlock block : blocks) {
                if (block.lastUsedTime < lru.lastUsedTime) {
                    lru = block;
                }
            }
            return lru;
        }
    }

    public void initialize(int cSize, int assoc, int bSize) {
        cacheSize = cSize;
        blockSize = bSize;
        associativity = assoc;
        numSets = (cacheSize * 1024) / (blockSize * associativity);

        cache = new CacheSet[numSets];
        for (int i = 0; i < numSets; i++) {
            cache[i] = new CacheSet(associativity, blockSize);

        }
    }

    public static void read (String file) throws FileNotFoundException{
        Scanner sc = new Scanner(new File(file));
        while (sc.hasNextLine()) {
            String line = sc.nextLine();
            String[] strings = line.split(" ");
            String type = strings[0];
            int address = Integer.parseInt(strings[1].substring(2), 16);
            int size = Integer.parseInt(strings[2]);
            //int value = null;
            if (strings.length > 3) {
                //System.out.println(strings[3]);
                //value = Long.parseInt(strings[3], 16);
            }
            if (type.equals("store")) {
                store(address, size, strings[3]);
            }
            else if (type.equals("load")) {
                load(address, size);
            }
        }
        sc.close();
    }

    public static void load(int address, int size) {
        AccessResult result = performAccess(address, "load", null);
        if (result.hit) {
            System.out.printf("load 0x%x hit ", address);
        }
        else {
            System.out.printf("load 0x%x miss ", address);
        }

        int startOfBlock = address/blockSize * blockSize; 
        int offset = address - startOfBlock;
        //System.out.printf("size: %d\n", size);
        for(int i = 0;i < size; i++){
            byte b = result.loadedBlock.data[i+offset];
            //System.out.printf("i + offset bit index: %d string: %02x\n", i/2+offset, b & 0xFF);

            System.out.printf("%02x", b);
        }
        System.out.printf("\n");
    }

    public static void store(int address, int size, String value) {
        AccessResult result = performAccess(address, "store", value);
        if (result.hit) {
            System.out.printf("store 0x%x hit\n", address);
        }
        else {
            System.out.printf("store 0x%x miss\n", address);
        }
    }

    public static class AccessResult {
        boolean hit;
        CacheBlock loadedBlock;

        public AccessResult(boolean hit, CacheBlock value) {
            this.hit = hit;
            this.loadedBlock = value;
        }
    }

    public static AccessResult performAccess(int address, String type, String value) {
        int offsetBits = log2(blockSize);
        int indexBits = log2(numSets);
        int index = (address >> offsetBits) & ((1 << indexBits) - 1);
        long tag = address >> (offsetBits + indexBits);
        int startOfBlock = address/blockSize * blockSize; 
        int offset = address - startOfBlock;
        CacheSet set = cache[index];
        CacheBlock targetBlock = null;
        boolean hit = false;
        int loadedValue = 0;

        for (CacheBlock block : set.blocks) {
            if (block.valid && block.tag == tag) {
                targetBlock = block;
                hit = true;
                if (type.equals("load")) {
                    //loadedValue = ByteBuffer.wrap(block.data).getInt();
                }
                else if (type.equals("store")) {
                    block.dirty = true;
                    for(int i = 0; i < value.length(); i+=2){
                        //System.out.printf("i + offset bit index: %d string: %s\n", i/2 + offset, value.substring(i, i+2));
                        block.data[i/2 + offset] = (byte)Integer.parseInt(value.substring(i,i+2), 16);
                    }
                    /** 
                    if (value != null) {
                        ByteBuffer.wrap(block.data).putInt(value);
                    }*/

                }
                break;
            }
        }
        if (!hit) {
            targetBlock = set.findLRU();
            String clean = "dirty";
            if(!targetBlock.dirty)clean = "clean";
            if (targetBlock.valid) {
                //System.arraycopy(targetBlock.data, 0, mainMemory, targetBlock.frameStart, blockSize);
                for(int i = 0;i<blockSize; i++){
                    mainMemory[targetBlock.frameStart+i] = targetBlock.data[i];
                }
                System.out.printf("replacement 0x%x %s\n", targetBlock.frameStart, clean);
            }
            //System.arraycopy(mainMemory, address & ~((1 << offsetBits) - 1), targetBlock.data, 0, blockSize);
            for(int i = 0; i < blockSize; i++){
                targetBlock.data[i] = mainMemory[startOfBlock+i];
            }
            targetBlock.tag = tag;
            if(type.equals("store")){
                for(int i = 0; i < value.length(); i+=2){
                    //System.out.printf("i + offset bit index: %d string: %s\n", i/2+offset, value.substring(i, i+2));
                    targetBlock.data[i/2 + offset] = (byte)Integer.parseInt(value.substring(i,i+2), 16);
                }
            }
            targetBlock.frameStart = startOfBlock;
            targetBlock.valid = true;
            targetBlock.dirty = false;
            if(type.equals("store"))targetBlock.dirty = true;
            targetBlock.lastUsedTime = ++currentTime;
            //loadedValue = ByteBuffer.wrap(targetBlock.data).getInt();
        }
        else {
            targetBlock.lastUsedTime = ++currentTime;
        }
        return new AccessResult (hit, targetBlock);
    }

    public static void main(String[] args) throws FileNotFoundException{
        if (args.length != 4) {
            System.out.println("Error");
            return;
        }
        String file = args[0];
        int cSize = Integer.parseInt(args[1]);
        int assoc = Integer.parseInt(args[2]);
        int bSize = Integer.parseInt(args[3]);

        cachesim cacheSim1 = new cachesim();
        cacheSim1.initialize(cSize, assoc, bSize);
        cacheSim1.read(file);
    }

    public static int log2(int n) {
        int r=0;
        while (n != 1) {
            n >>= 1;
            r++;
        }
        return r;
    }
}